File: src/main/java/org/moinex/util/SuggestionsHandlerHelper.java
/*
 * Filename: SuggestionsHelper.java
 * Created on: March 15, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.util;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;
import javafx.beans.value.ChangeListener;
import javafx.scene.control.Label;
import javafx.scene.control.ListCell;
import javafx.scene.control.ListView;
import javafx.scene.control.TextField;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;
import javafx.stage.Popup;
import lombok.Setter;

public class SuggestionsHandlerHelper<T> {
    private final ListView<T> suggestionListView;
    private final Popup suggestionsPopup;
    private final Function<T, String> displayFunction;
    private final Function<T, String> filterFunction;
    private final Consumer<T> onSelectCallback;
    private final TextField componentField;
    private ChangeListener<String> componentFieldListener;
    @Setter private List<T> suggestions = new ArrayList<>();

    public SuggestionsHandlerHelper(
            TextField componentField,
            Function<T, String> filterFunction,
            Function<T, String> displayFunction,
            Consumer<T> onSelectCallback) {
        this.suggestionListView = new ListView<>();
        this.suggestionsPopup = new Popup();

        this.componentField = componentField;
        this.filterFunction = filterFunction;
        this.displayFunction = displayFunction;
        this.onSelectCallback = onSelectCallback;

        configureListeners();
        configureSuggestionsListView();
        configureSuggestionsPopup();
    }

    public void disable() {
        componentField.textProperty().removeListener(componentFieldListener);
    }

    public void enable() {
        componentField.textProperty().addListener(componentFieldListener);
    }

    private void configureListeners() {
        // Store the listener in a variable to be able to disable and enable it
        // when needed
        componentFieldListener =
                (observable, oldValue, newValue) -> {
                    if (newValue.isBlank()) {
                        suggestionsPopup.hide();
                        return;
                    }

                    suggestionListView.getItems().clear();
                    List<T> filteredSuggestions =
                            suggestions.stream()
                                    .filter(
                                            item ->
                                                    filterFunction
                                                            .apply(item)
                                                            .toLowerCase()
                                                            .contains(newValue.toLowerCase()))
                                    .toList();

                    if (filteredSuggestions.size() > Constants.SUGGESTIONS_MAX_ITEMS) {
                        filteredSuggestions =
                                filteredSuggestions.subList(0, Constants.SUGGESTIONS_MAX_ITEMS);
                    }

                    suggestionListView.getItems().addAll(filteredSuggestions);

                    if (!filteredSuggestions.isEmpty()) {
                        adjustPopupWidth();
                        adjustPopupHeight();
                        showPopup();
                    } else {
                        suggestionsPopup.hide();
                    }
                };
    }

    private void configureSuggestionsListView() {
        suggestionListView.setCellFactory(
                param ->
                        new ListCell<>() {
                            @Override
                            protected void updateItem(T item, boolean empty) {
                                super.updateItem(item, empty);
                                if (empty || item == null) {
                                    setText(null);
                                } else {
                                    VBox cellContent = new VBox();
                                    cellContent.setSpacing(2);

                                    Label descriptionLabel = new Label(displayFunction.apply(item));

                                    cellContent.getChildren().addAll(descriptionLabel);
                                    setGraphic(cellContent);
                                }
                            }
                        });

        suggestionListView.setPrefWidth(Region.USE_COMPUTED_SIZE);
        suggestionListView.setPrefHeight(Region.USE_COMPUTED_SIZE);

        // By default, the SPACE key is used to select an item in the ListView.
        // This behavior is not desired in this case, so the event is consumed
        suggestionListView.addEventFilter(
                KeyEvent.KEY_PRESSED,
                event -> {
                    if (event.getCode() == KeyCode.SPACE) {
                        event.consume(); // Do not propagate the event
                    }
                });

        suggestionListView
                .getSelectionModel()
                .selectedItemProperty()
                .addListener(
                        (observable, oldValue, newValue) -> {
                            if (newValue != null) {
                                onSelectCallback.accept(newValue);
                                suggestionsPopup.hide();
                            }
                        });
    }

    private void configureSuggestionsPopup() {
        suggestionsPopup.setAutoHide(true);
        suggestionsPopup.setHideOnEscape(true);
        suggestionsPopup.getContent().add(suggestionListView);
    }

    private void adjustPopupWidth() {
        suggestionListView.setPrefWidth(componentField.getWidth());
    }

    private void adjustPopupHeight() {
        int itemCount =
                Math.min(suggestionListView.getItems().size(), Constants.SUGGESTIONS_MAX_ITEMS);
        double totalHeight = itemCount * 45.0;
        suggestionListView.setPrefHeight(totalHeight);
    }

    private void showPopup() {
        suggestionsPopup.show(
                componentField,
                componentField.localToScene(0, 0).getX()
                        + componentField.getScene().getWindow().getX()
                        + componentField.getScene().getX(),
                componentField.localToScene(0, 0).getY()
                        + componentField.getScene().getWindow().getY()
                        + componentField.getScene().getY()
                        + componentField.getHeight());
    }
}

File: src/main/kotlin/org/moinex/util/SuggestionsHandlerHelper.kt
/*
 * Filename: SuggestionsHandlerHelper.kt (original filename: SuggestionsHandlerHelper.java)
 * Created on: March 15, 2025
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrated to Kotlin on 18/03/2026
 */

package org.moinex.util

import javafx.beans.value.ChangeListener
import javafx.scene.control.Label
import javafx.scene.control.ListCell
import javafx.scene.control.ListView
import javafx.scene.control.TextField
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.Region
import javafx.scene.layout.VBox
import javafx.stage.Popup
import org.moinex.constants.Constants
import java.util.function.Consumer
import java.util.function.Function

class SuggestionsHandlerHelper<T>(
    private val componentField: TextField,
    private val filterFunction: Function<T, String>, // TODO: Após migração para kotlin, utilizar (T) -> String
    private val displayFunction: Function<T, String>, // TODO: Após migração para kotlin, utilizar (T) -> String
    private val onSelectCallback: Consumer<T>, // TODO: Após migração para kotlin, utilizar (T) -> Unit
) {
    private val suggestionListView = ListView<T>()
    private val suggestionsPopup = Popup()
    private lateinit var componentFieldListener: ChangeListener<String>

    var suggestions: List<T> = emptyList()

    init {
        configureListeners()
        configureSuggestionsListView()
        configureSuggestionsPopup()
    }

    fun disable() {
        componentField.textProperty().removeListener(componentFieldListener)
    }

    fun enable() {
        componentField.textProperty().addListener(componentFieldListener)
    }

    private fun configureListeners() {
        componentFieldListener =
            ChangeListener { _, _, newValue ->
                if (newValue.isBlank()) {
                    suggestionsPopup.hide()
                    return@ChangeListener
                }

                suggestionListView.items.clear()

                val filteredSuggestions =
                    suggestions
                        .filter { item ->
                            filterFunction
                                .apply(item)
                                .lowercase()
                                .contains(newValue.lowercase())
                        }.take(Constants.SUGGESTIONS_MAX_ITEMS)

                suggestionListView.items.addAll(filteredSuggestions)

                if (filteredSuggestions.isNotEmpty()) {
                    adjustPopupWidth()
                    adjustPopupHeight()
                    showPopup()
                } else {
                    suggestionsPopup.hide()
                }
            }
    }

    private fun configureSuggestionsListView() {
        suggestionListView.apply {
            setCellFactory {
                object : ListCell<T>() {
                    override fun updateItem(
                        item: T?,
                        empty: Boolean,
                    ) {
                        super.updateItem(item, empty)

                        graphic =
                            if (empty || item == null) {
                                null
                            } else {
                                VBox().apply {
                                    spacing = 2.0
                                    children.add(Label(displayFunction.apply(item)))
                                }
                            }
                    }
                }
            }

            prefWidth = Region.USE_COMPUTED_SIZE
            prefHeight = Region.USE_COMPUTED_SIZE

            addEventFilter(KeyEvent.KEY_PRESSED) { event ->
                if (event.code == KeyCode.SPACE) {
                    event.consume()
                }
            }

            selectionModel.selectedItemProperty().addListener { _, _, newValue ->
                newValue?.let {
                    onSelectCallback.accept(it)
                    suggestionsPopup.hide()
                }
            }
        }
    }

    private fun configureSuggestionsPopup() {
        suggestionsPopup.apply {
            isAutoHide = true
            isHideOnEscape = true
            content.add(suggestionListView)
        }
    }

    private fun adjustPopupWidth() {
        suggestionListView.prefWidth = componentField.width
    }

    private fun adjustPopupHeight() {
        val itemCount = suggestionListView.items.size.coerceAtMost(Constants.SUGGESTIONS_MAX_ITEMS)
        suggestionListView.prefHeight = itemCount * 45.0
    }

    private fun showPopup() {
        val scene = componentField.scene ?: return
        val window = scene.window ?: return

        suggestionsPopup.show(
            componentField,
            componentField.localToScene(0.0, 0.0).x + window.x + scene.x,
            componentField.localToScene(0.0, 0.0).y + window.y + scene.y + componentField.height,
        )
    }
}
